x/incentive/keeper/rewards.go

```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
 	k.SetHardLiquidityProviderClaim(ctx, claim)
 }
 
+// SynchronizeHardDelegatorRewards updates the claim object by adding any accumulated rewards
+func (k Keeper) SynchronizeHardDelegatorRewards(ctx sdk.Context, delegator sdk.AccAddress) {
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, delegator)
+	if !found {
+		return
+	}
+
+	delagatorFactor, found := k.GetHardDelegatorRewardFactor(ctx, UKAVA_DENOM)
+	if !found {
+		return
+	}
+
+	delegatorIndex, hasDelegatorRewardIndex := claim.HasDelegatorRewardIndex(UKAVA_DENOM)
+	if !hasDelegatorRewardIndex {
+		return
+	}
+
+	userRewardFactor := claim.DelegatorRewardIndexes[delegatorIndex].RewardFactor
+	rewardsAccumulatedFactor := delagatorFactor.Sub(userRewardFactor)
+	if rewardsAccumulatedFactor.IsZero() {
+		return
+	}
+	claim.DelegatorRewardIndexes[delegatorIndex].RewardFactor = delagatorFactor
+
+	var totalDelegated sdk.Dec
+
+	// TODO: set reasonable max limit on delegation iteration
+	maxUInt := ^uint16(0)
+	delegations := k.stakingKeeper.GetDelegatorDelegations(ctx, delegator, maxUInt)
+	for _, delegation := range delegations {
+		validator, found := k.stakingKeeper.GetValidator(ctx, delegation.GetValidatorAddr())
+		if !found {
+			continue
+		}
+
+		// Delegators don't accumulate rewards if their validator is unbonded/slashed
+		if validator.GetStatus() != sdk.Bonded {
+			continue
+		}
+
+		if validator.GetTokens().IsZero() {
+			continue
+		}
+
+		sharesToTokens := sdk.NewDecFromInt(validator.GetTokens()).Quo(validator.GetDelegatorShares())
+		delegatedTokens := sharesToTokens.Mul(delegation.GetShares())
+		if delegatedTokens.IsZero() || delegatedTokens.IsNegative() {
+			continue
+		}
+
+		totalDelegated = totalDelegated.Add(delegatedTokens)
+		// TODO: do we need to consider respective share of validator tokens?
```

Comment:
don't think so. Just calculating the number of bonded tokens that this delegator accounts for is sufficient.

Reply:
Removed TODO in https://github.com/Kava-Labs/kava/pull/776/commits/4ec5fe20e9d8cde0a46fda55dcf02e42e3afd732